### PR TITLE
reduce amount allocation for single number decode

### PR DESCRIPTION
**Pull request by jetersen**
Reason for change:
Almost 5x saving for single decode
After further changes by adding a optimized splitter using `Span<char>` it is only a 2x saving.
By adding a specific `GetNumberFrom` that handles a single number we avoid creating unnecessary array for `long[]` and can simply pass back `long`

I think it would greatly benefit to also actually have a `GenerateHashFrom` that takes single int/long that would mean single encode decode could be 100% allocation free.

```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1620 (21H2)
AMD Ryzen 9 3950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK=6.0.103
  Job-RCMHEX : .NET 6.0.3 (6.0.322.12309), X64 RyuJIT

|               Method |        Job |            Runtime |       Mean |     Error |    StdDev | Code Size |  Gen 0 | Allocated |
|--------------------- |----------- |------------------- |-----------:|----------:|----------:|----------:|-------:|----------:|
|        DecodeNumbers | Job-RCMHEX |           .NET 6.0 |   864.4 ns |   3.99 ns |   3.74 ns |   1,359 B | 0.0181 |     152 B |
|   DecodeSingleNumber | Job-RCMHEX |           .NET 6.0 |   799.2 ns |   5.65 ns |   5.28 ns |     190 B | 0.0038 |      32 B |
```

### CODE REVIEW

**Comment by jetersen** [comment]
I may have missed the purpose of the `_guards` split in the GetNumbersFrom and whether it applies to single.

**Comment by KeterSCP** [comment]
The following code throws `NoResultException` in your PR if no _guards were used:

```c#
var hashids = new HashidsNet.Hashids("salt", minHashLength: 4);

var hash = hashids.Encode(123);
Console.WriteLine(hash);

var decoded = hashids.DecodeSingle(hash); // <-- Exception here
Console.WriteLine(decoded);
```

**Comment by jetersen** [comment]
Thanks I'll take a look.

**Comment by jetersen** [comment]
I think I managed to find a way to actually improve Decode for multiple numbers without allocating string arrays for guards 😄

**Comment by jetersen** [comment]
@KeterSCP [bc5f785](https://github.com/ullmark/hashids.net/pull/65/commits/bc5f7857541791060d67476b6f8f3fc4ff1fc70b) could also apply to multiple decoding.

**Comment by KeterSCP** [comment]
> The following code throws `NoResultException` in your PR if no _guards were used:
> 
> ```cs
> var hashids = new HashidsNet.Hashids("salt", minHashLength: 4);
> 
> var hash = hashids.Encode(123);
> Console.WriteLine(hash);
> 
> var decoded = hashids.DecodeSingle(hash); // <-- Exception here
> Console.WriteLine(decoded);
> ```

@jetersen this still throws with `minHashLength=5`

**Comment by jetersen** [comment]
Damn empty entries 🤣🤣🤣🤣

**Comment by jetersen** [comment]
```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1620 (21H2)
AMD Ryzen 9 3950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK=6.0.103
  [Host]     : .NET 6.0.3 (6.0.322.12309), X64 RyuJIT
  Job-AEAOHU : .NET 6.0.3 (6.0.322.12309), X64 RyuJIT

OutlierMode=DontRemove  Runtime=.NET 6.0  MemoryRandomization=True  
```

Before: [a7478cd](https://github.com/ullmark/hashids.net/pull/65/commits/a7478cd1b8e8dbe02dde8c7f609b5d3d7388cd97)
```
|               Method |       Mean |    Error |   StdDev | Code Size |  Gen 0 | Allocated |
|--------------------- |-----------:|---------:|---------:|----------:|-------:|----------:|
|        RoundtripInts | 3,785.7 ns | 18.53 ns | 17.33 ns |      50 B | 0.0610 |     512 B |
|       RoundtripLongs | 4,287.0 ns | 18.31 ns | 17.13 ns |   3,330 B | 0.0610 |     512 B |
|         RoundtripHex | 3,725.4 ns | 22.88 ns | 21.41 ns |      49 B | 0.1602 |   1,344 B |
|         SingleNumber |   844.0 ns |  3.69 ns |  3.45 ns |     145 B | 0.0076 |      64 B |
|        DecodeNumbers |   862.9 ns |  4.09 ns |  3.82 ns |   1,359 B | 0.0181 |     152 B |
|      RoundtripSingle | 1,255.6 ns |  8.97 ns |  8.39 ns |     304 B | 0.0076 |      64 B |
| SingleNumberAsParams |   862.7 ns |  4.93 ns |  4.61 ns |   2,040 B | 0.0191 |     160 B |
```

After: [a7478cd](https://github.com/ullmark/hashids.net/pull/65/commits/a7478cd1b8e8dbe02dde8c7f609b5d3d7388cd97)
```
|               Method |       Mean |    Error |   StdDev | Code Size |  Gen 0 | Allocated |
|--------------------- |-----------:|---------:|---------:|----------:|-------:|----------:|
|        RoundtripInts | 3,827.1 ns | 25.90 ns | 24.22 ns |      50 B | 0.0305 |     264 B |
|       RoundtripLongs | 4,356.8 ns | 34.19 ns | 31.98 ns |   3,612 B | 0.0229 |     224 B |
|         RoundtripHex | 3,817.9 ns | 13.13 ns | 12.28 ns |      49 B | 0.1335 |   1,128 B |
|         SingleNumber |   846.4 ns |  5.44 ns |  5.09 ns |     145 B | 0.0076 |      64 B |
|        DecodeNumbers |   864.6 ns |  3.67 ns |  3.43 ns |   1,641 B | 0.0076 |      64 B |
|      RoundtripSingle | 1,248.2 ns |  7.87 ns |  7.36 ns |     304 B | 0.0076 |      64 B |
| SingleNumberAsParams |   855.6 ns |  3.93 ns |  3.67 ns |   2,040 B | 0.0191 |     160 B |
```

**Comment by KeterSCP** [comment]
New test for `long.MaxValue` fails in this PR

image

**Comment by KeterSCP** [comment]
Hmm, the following code produces different exception for `main` and PR branches:

```c#
var hashids = new HashidsNet.Hashids("salt", minHashLength: 0, alphabet: "0123456789ABCDEF");

var hash = hashids.EncodeLong(long.MaxValue);
Console.WriteLine(hash);

var decoded = hashids.DecodeSingleLong(hash + "1"); // HashidsNet.NoResultException on main branch and HashidsNet.MultipleResultsException on PR branch
Console.WriteLine(decoded);
```

**Comment by jetersen** [comment]
@KeterSCP That's true because I decided to check for separators ahead of time. Previous code was doing after the decode had returned multiple numbers.

So that's a toss up and this behavior was only added in 1.5.0 so which behavior is more correct?

**Comment by KeterSCP** [comment]
> @KeterSCP That's true because I decided to check for separators ahead of time. Previous code was doing after the decode had returned multiple numbers.
> 
> So that's a toss up and this behavior was only added in 1.5.0 so which behavior is more correct?
> 
> See #57

I've checked the 1.4.1 version and following sample produces array of 0 length:

```c#
var hashids = new HashidsNet.Hashids("salt", minHashLength: 0, alphabet: "0123456789ABCDEF");

var hash = hashids.EncodeLong(1000);
Console.WriteLine(hash);

var decodedArray = hashids.DecodeLong(hash + "1");
Console.WriteLine(decodedArray.Length); // <-- 0 (No results decoded)
```

While on the PR branch this gives a MultipleResultsException.
Generally speaking, I support such perf changes as long as they mirror old behavior, otherwise, it would introduce a breaking change.

**Comment by manigandham** [comment]
@jetersen 

Can you please rebase on master to use the latest tests?

Also agree with @KeterSCP that the behavior shouldn't change. If there are no results decoded then it should not be returning a MultipleEntries exception. Is your code finding multiple integers in that hash?

**Comment by jetersen** [comment]
I took your example @KeterSCP 

You specifically added a one and with your setup tat is one of the separators:

image

So how can `DecodeSingle` be wrong?

**Comment by jetersen** [comment]
Rebased

**Comment by manigandham** [comment]
> Rebased

The log doesn't show the changes. You might need to pull down the changes on `master` first before rebasing...

**Comment by jetersen** [comment]
Rebased successfully 😅 I forgot to fetch upstream before rebasing.

**Comment by KeterSCP** [comment]
> I took your example @KeterSCP 
> 
> You specifically added a `1` to decode and with your setup that is one of the separators:
> 
> image
> 
> So how can `DecodeSingle` be wrong?
> 
> I can remove the check for separators but than I should also remove the `MultipleEntriesException` cause there is no way to detect other than to look for separators after the result does not match the `GenerateHashFrom` equals check.
> 
> I tried to keep the behavior of `DecodeSingleLong` as it was coded in 1.5.0:
> https://github.com/ullmark/hashids.net/blob/b2c78773d05fcf8c9edc643f87db31e98554f2f0/src/Hashids.net/Hashids.cs#L180-L181

Well, i've just checked how it behaves for invalid hashes and compared how it did before. If such change is ok for the author of library, then i suppose it can be merged. Code looks good.
cc: @ullmark

**Comment by jetersen** [comment]
I created 535df3d as another way. This would result in NoResultException when trying to parse multiple when using `DecodeSingle`

File: src/Hashids.net/Hashids.cs [hunk 1]
```diff
@@ -201,25 +200,23 @@ public bool TryDecodeSingleLong(string hash, out long id)
...
         }
 
         /// <inheritdoc />
         public virtual bool TryDecodeSingle(string hash, out int id)
         {
-            var numbers = GetNumbersFrom(hash);
+            var number = GetNumberFrom(hash);
 
-            if (numbers.Length == 1)
+            if (number > 0)
```
Comment (Berezhnyk) [inline src/Hashids.net/Hashids.cs:217]: Should it be number >= 0 for the case when id is 0?

Comment (jetersen) [reply src/Hashids.net/Hashids.cs:217]: Yes 👍

File: src/Hashids.net/Hashids.cs [hunk 2]
```diff
@@ -274,6 +271,103 @@ public virtual string DecodeHex(string hash)
...
+                stringBuilder.Append(alphabet[..halfLength]);
+#endif
+
+                var excess = stringBuilder.Length - _minHashLength;
+                if (excess > 0)
+                {
+                    stringBuilder.Remove(0, excess / 2);
+                    stringBuilder.Remove(_minHashLength, stringBuilder.Length - _minHashLength);
+                }
+            }
```
Comment (jetersen) [inline src/Hashids.net/Hashids.cs:355]: I was unsure how to implement the remove excess that basically seems to reverse the alphabet but again unsure how to appropriately do that with pure spans and limited space for the result.

The ValueStringBuilder would be great.

**Comment by jetersen** [comment]
After 99a41e4, 395a2d6 and 7567d90:
```
|               Method |       Mean |    Error |   StdDev |  Gen 0 | Code Size | Allocated |
|--------------------- |-----------:|---------:|---------:|-------:|----------:|----------:|
|        RoundtripInts | 3,835.8 ns | 11.98 ns | 11.21 ns | 0.0229 |      50 B |     200 B |
|       RoundtripLongs | 4,389.7 ns | 16.26 ns | 15.21 ns | 0.0153 |     862 B |     136 B |
|         RoundtripHex | 3,843.8 ns |  9.92 ns |  9.28 ns | 0.1259 |      49 B |   1,064 B |
|         SingleNumber |   875.7 ns |  1.76 ns |  1.65 ns | 0.0076 |     196 B |      64 B |
|        DecodeNumbers |   875.4 ns |  8.87 ns |  8.30 ns | 0.0038 |     501 B |      32 B |
|   DecodeSingleNumber |   814.0 ns |  2.66 ns |  2.49 ns |      - |     118 B |         - |
|      RoundtripSingle | 1,272.6 ns | 12.14 ns | 11.36 ns | 0.0038 |     283 B |      32 B |
| SingleNumberAsParams |   911.1 ns |  4.11 ns |  3.84 ns | 0.0191 |     450 B |     160 B |
```

I would like to see if it possible to get them to return `ReadOnlySpan<char>` to have return option that is only on the stack.

But `GenerateHashFrom` now allows us to potentially return

File: src/Hashids.net/Hashids.cs [hunk 3]
```diff
@@ -138,14 +139,25 @@ public Hashids() : this(salt: string.Empty, minHashLength: 0, alphabet: DEFAULT_
         /// </summary>
         /// <param name="number">the number</param>
         /// <returns>the hashed string</returns>
-        public string EncodeLong(long number) => GenerateHashFrom(stackalloc[] { number });
+        public string EncodeLong(long number)
+        {
+            Span<char> result = stackalloc char[20];
```
Comment (KeterSCP) [inline src/Hashids.net/Hashids.cs:144]: Why there is a limit of 20 chars? Now it throws if you try to encode value with `minHashLength` parameter set to >20, however previously it was possible to get string of any length

Comment (jetersen) [reply src/Hashids.net/Hashids.cs:144]: I thought the max length for a long was 19 plus 1 for lottery hence 20.

We should have a test that validate any length.

Comment (jetersen) [reply src/Hashids.net/Hashids.cs:144]: Also potentially any length does not make sense when it comes to max url length or other serialization issues or memory optimization.

Comment (KeterSCP) [reply src/Hashids.net/Hashids.cs:144]: Don't confuse input number length (e.g `var num = 1234` has length of 4) and output hash length (which can differ)

**Comment by manigandham** [comment]
hey @jetersen 

The fundamental problem here is that `DecodeSingle` is still not truthful about the error. Previously it would return `MultipleResultsException` even if there was nothing decoded, but now the code never returns `MultipleResultsException` even if there are, effectively pointing users to the wrong error.

As nice as it is to have improved performance, I don't think it should come at the cost of correct results. Is there a way to fix this?

**Comment by jetersen** [comment]
@manigandham As you can see there was an argument about this

Whether it returns `NoResultException` or `MultipleEntriesException` to me is irrelevant.

But as you can see I suggest I can revert that change in 535df3d

Before 535df3d it is possible to detect multiple entries but apparently the way I do that was not appreciated.

I was looking for any Separator characters, and throw `MultipleEntriesException`, see https://github.com/ullmark/hashids.net/pull/65#issuecomment-1088045441 and see https://github.com/ullmark/hashids.net/pull/65#issuecomment-1088413534

Again `MultipleEntriesException` was only added in recent version as well. So in my opinion we could do either. Revert 535df3d to throw `MultipleEntriesException` or keep 535df3d thereby throwing `NoResultException`

What difference does it make `MultipleEntriesException` is also the same as `NoResultException` both exception is a non result situation.

**Comment by manigandham** [comment]
> What difference does it make MultipleEntriesException is the same as NoResultException both exception is a non result situation.

An exception being thrown signals that there's an issue with the operation, but the actual exception type describes *why* there was an issue. I think it's very important for the user to know so they understand why the result didn't decode (whether there's multiple numbers or none).

I'll leave it to @ullmark to decide how to proceed.

**Comment by ullmark** [comment]
@manigandham If the performance increases of this library is important for the ones who uses it, and this looks good we should merge this. 

tbh; it uses some syntax I've never seen or used (or had any need to) in C#, so @jetersen you are welcome to come in and maintain if you like. :) 

Re. the errors, if it's not possible to detect multiple entries in a good way, just throw a "NoResultException" since that is still true, given the alphabet + salt, and how you use the library, we weren't able to get a result. 🤷‍♂️

**Comment by manigandham** [comment]
@jetersen if you want to change the exception to say `NoResultException` then I'll merge in and do the next release.

**Comment by jetersen** [comment]
@manigandham it is already throwing `NoResultException`, see the tests in `GeneralTests.cs` :)

**Comment by jetersen** [comment]
@ullmark I would be happy to help maintain and review PRs 😄

**Review by manigandham — APPROVED** [review]

**Comment by manigandham** [comment]
v1.6.0 released https://www.nuget.org/packages/Hashids.net/1.6.0

**Pull request merged**